pkg/query/recursive.go

```diff
@@ -0,0 +1,184 @@
+package query
+
+import (
+	"fmt"
+)
+
+const defaultMaxRecursionDepth = 50
+
+// RecursiveIterator is the root controller that manages iterative deepening for recursive schemas.
+// It wraps an iterator tree that contains RecursiveSentinel sentinels, and executes the tree
+// repeatedly with increasing depth until a fixed point is reached or max depth is exceeded.
+type RecursiveIterator struct {
+	templateTree Iterator
+	sentinels    []*RecursiveSentinel
+}
+
+// NewRecursiveIterator creates a new recursive iterator controller
+func NewRecursiveIterator(templateTree Iterator, sentinels []*RecursiveSentinel) *RecursiveIterator {
+	return &RecursiveIterator{
+		templateTree: templateTree,
+		sentinels:    sentinels,
+	}
+}
+
+// CheckImpl implements iterative deepening for Check operations
+func (r *RecursiveIterator) CheckImpl(ctx *Context, resources []Object, subject ObjectAndRelation) (PathSeq, error) {
+	return r.iterativeDeepening(ctx, func(ctx *Context, tree Iterator) (PathSeq, error) {
+		return ctx.Check(tree, resources, subject)
+	})
+}
+
+// IterSubjectsImpl implements iterative deepening for IterSubjects operations
+func (r *RecursiveIterator) IterSubjectsImpl(ctx *Context, resource Object) (PathSeq, error) {
+	return r.iterativeDeepening(ctx, func(ctx *Context, tree Iterator) (PathSeq, error) {
+		return ctx.IterSubjects(tree, resource)
+	})
+}
+
+// IterResourcesImpl implements iterative deepening for IterResources operations
+func (r *RecursiveIterator) IterResourcesImpl(ctx *Context, subject ObjectAndRelation) (PathSeq, error) {
+	return r.iterativeDeepening(ctx, func(ctx *Context, tree Iterator) (PathSeq, error) {
+		return ctx.IterResources(tree, subject)
+	})
+}
+
+// iterativeDeepening executes the core iterative deepening algorithm
+// It yields results directly, always running to maxDepth to find all valid paths
```

Comment:
Does this mean that it's eager when it's executed?

Reply:
It just creates a new iterator -- the evaluation of which is in the `execute` callback, below, which happens in the greater iteration step... so these only get deepened as needed